M2/Macaulay2/packages/Tableaux/SkewTableaux.m2

```diff
@@ -0,0 +1,578 @@
+    skewTableau(lam,new Partition from {},theList)
+    )
+skewTableau (Partition,Partition) := (lam,mu) -> (
+    (lam',mu') := standardize (lam,mu);
+    numBoxesNeeded := sum for i from 0 to #lam'-1 list abs(lam'#i-mu'#i);
+    
+    skewTableau(lam,mu,toList(numBoxesNeeded:""))
+    )
+skewTableau Partition := lam -> (
+    numBoxesNeeded := sum for i from 0 to #lam-1 list abs(lam#i);
+    
+    skewTableau(lam, new Partition from {}, toList(numBoxesNeeded:""))
+    )
+
+skewShape = method(TypicalValue => Sequence)
+skewShape SkewTableau := T -> (
+    (T#"outerShape", T#"innerShape")
+    )
+
+truncate (Partition,Partition) := theInput -> (
+    (lam,mu) := theInput#1;
+
+    lamNumTrailingZeros := # for i from 1 to #lam list (if lam#-i == 0 then 1 else break);
+    muNumTrailingZeros := # for i from 1 to #mu list (if mu#-i == 0 then 1 else break);
+
+    lamShortened := (toList lam)_(toList(0..(#lam-1-lamNumTrailingZeros)));
+    muShortened := (toList mu)_(toList(0..(#mu-1-muNumTrailingZeros)));
+
+    (new Partition from lamShortened, new Partition from muShortened)
+    )
+
+pad (Partition,Partition) := (lam,mu) -> (
+    maxLength := max(#lam,#mu);
+
+    lamPadded := toList(lam)|toList((maxLength - #lam):0);
+    muPadded := toList(mu)|toList((maxLength - #mu):0);
+    
+    (new Partition from lamPadded, new Partition from muPadded)
+    )
+
+standardize = method(TypicalValue => Sequence)
+standardize (Partition,Partition) := (lam,mu) -> pad truncate (lam,mu)
+
+alignToZero = method(TypicalValue => SkewTableau)
+alignToZero (Partition,Partition) := (lam,mu) -> (
+    (lam,mu) = standardize (lam,mu);
+    minPart := min(toList mu | toList lam);
+
+    lamAligned := for thePart in lam list thePart - minPart;
+    muAligned := for thePart in mu list thePart - minPart;
+
+    (new Partition from lamAligned, new Partition from muAligned)
+    )
+alignToZero SkewTableau := T -> (
+    (lamAligned,muAligned) := alignToZero standardize skewShape T;
+    skewTableau(lamAligned, muAligned, entries T)
+    )
+
+hasNegativeRow = method(TypicalValue => Boolean)
+hasNegativeRow SkewTableau := T -> (
+    (lam,mu) := standardize skewShape T;
+    
+    any(0..(#lam-1), i -> mu#i > lam#i)
+    )
+
+-- swap parts when mu#i > lam#i
+normalizeNegativeRows = method(TypicalValue => SkewTableau)
+normalizeNegativeRows SkewTableau := T -> (
+    if not hasNegativeRow T then return T;
+    
+    (lam,mu) := standardize skewShape T;
+
+    lam' := new Partition from for i from 0 to #lam-1 list max(lam#i,mu#i);
+    mu' := new Partition from for i from 0 to #lam-1 list min(lam#i,mu#i);
+    
+    skewTableau(lam',mu',entries T)
+    )
+
+listNegativeRows = method(TypicalValue => List)
+listNegativeRows SkewTableau := T -> (
+    (lam,mu) := standardize skewShape T;
+
+    for i from 0 to #lam-1 list (mu#i > lam#i)
+    )
+
+-- retrieving rows and columns
+
+numcols SkewTableau := T -> (
+    (lam,mu) := standardize skewShape normalizeNegativeRows T;
```

Comment:
Would it make sense to use caching for `standardize skewShape normalizeNegativeRows T`? It seems to be called very frequently.